### ISSUE TITLE
canary-on-1.16 block tests flaky

### ISSUE THREAD

**Issue by msau42** [kubernetes-csi/csi-driver-host-path#119]
https://k8s-testgrid.appspot.com/sig-storage-csi-ci#canary-on-1.16

Oddly enough, canary-on-master tests seem fine.

@jsafrane @pohly could you help take a look?

/kind failing-test

**Comment by msau42** [comment]
Its flaky across all jobs, not just this.

I think it's related to running on kind, where hostpath volumes go to the host. There may be races due another test grabbing the same device

**Comment by msau42** [comment]
Although if that were the case, then intree parallel tests should be flaky too? Maybe it's a probability game since there a lot more tests running intree

**Comment by pohly** [comment]
> There may be races due another test grabbing the same device

Creating a loop device is done by kubelet and the CSI hostpath driver with https://github.com/kubernetes/kubernetes/blob/8a09460c2f7ba8f6acd8a6fb7603ed3ac4805eb6/pkg/volume/util/volumepathhandler/volume_path_handler_linux.go#L96-L112.

The command invocation itself (= "losetup -f --show") shouldn't be racy. Let's look at its implementation:
```
openat(AT_FDCWD, "/dev/loop-control", O_RDWR|O_CLOEXEC) = 3
ioctl(3, LOOP_CTL_GET_FREE)             = 0
close(3)                                = 0
lstat("/tmp", {st_mode=S_IFDIR|S_ISVTX|0777, st_size=260, ...}) = 0
lstat("/tmp/test", {st_mode=S_IFREG|0644, st_size=1073741824, ...}) = 0
openat(AT_FDCWD, "/tmp/test", O_RDWR|O_CLOEXEC) = 3
openat(AT_FDCWD, "/dev/loop0", O_RDWR|O_CLOEXEC) = 4
ioctl(4, LOOP_SET_FD, 3)                = 0
ioctl(4, LOOP_SET_STATUS64, {lo_offset=0, lo_number=0, lo_flags=0, lo_file_name="/tmp/test", ...}) = 0
```

`losetup` first determines the next available loop device, then opens it and configures it. At first glance this looks like it might be racy, but that depends on how `LOOP_SET_FD` and `losetup` behave when someone else grabs the device in parallel.

I tested that by running `losetup` under `gdb` and pausing it in the `LOOP_SET_FD` ioctl, then taking the loop device with some other `losetup` call. The first invocation recovered gracefully from that by retrying with another loop device.

So my conclusion is that we don't have a race around loop device creation.

Perhaps KinD containers have the same issue that we also had in the CSI hostpath container where `/dev` is a static copy of the host and then new loop devices don't show up?

https://github.com/kubernetes/kubernetes/blob/8a09460c2f7ba8f6acd8a6fb7603ed3ac4805eb6/pkg/volume/util/volumepathhandler/volume_path_handler_linux.go#L96-L112

Indeed, when I just ran the CSI hostpath prow tests on my development machine after using all existing loop devices, the blockvolume tests fail because `losetup` fails:
```
Jan 13 08:31:51.654: INFO: At 2020-01-13 08:27:28 +0100 CET - event for security-context-6ade56c3-0fd9-4080-a3b5-e46a57584a4c: {kubelet csi-prow-worker} FailedMapVolume: MapVolume.AttachFileDevice failed for volume "pvc-228ac164-bec1-43fe-a3d1-7671ef0549f6" : exit status 1
```

I've not found more about it in the logs, but I can reproduce the problem manually by creating loop devices with `docker exec f049e80f378a sh -c 'truncate -s 1G /tmp/testfile; losetup -f --show /tmp/testfile'` until eventually all the ones that existed when the KinD container was started are in use and the command fails with:
```
losetup: /tmp/testfile: failed to set up loop device: No such file or directory
```

I'll file an issue against KinD.
